test/parallel_api/algorithm/alg.sorting/sort_by_key_common.h

```diff
@@ -0,0 +1,297 @@
+    std::unordered_map<KeyT, Size> histogram;
+    std::for_each(keys_begin, keys_begin + n, [&histogram](const auto& key) { ++histogram[key]; });
+    auto first = oneapi::dpl::make_zip_iterator(keys_begin, vals_begin);
+    auto has_duplicates = [&histogram](const auto& pair) {return histogram[std::get<0>(pair)] > 1; };
+    auto new_last = std::remove_if(first, first + n, has_duplicates);
+    return std::distance(first, new_last);
+}
+
+template<typename KeyIt, typename ValIt, typename Size>
+void
+generate_data(KeyIt keys_begin, ValIt vals_begin, Size keys_n, Size vals_n, std::uint32_t seed)
+{
+    using KeyT = typename std::iterator_traits<KeyIt>::value_type;
+    using ValT = typename std::iterator_traits<ValIt>::value_type;
+    if constexpr (std::is_same_v<ValT, Particle>)
+    {
+        static_assert(std::is_same_v<KeyT, Particle::energy_type>);
+        std::default_random_engine gen{seed};
+        std::uniform_real_distribution<float> mass_dist(0.0, 1000.0);
+        std::uniform_real_distribution<float> velocity_dist(0.0, 1.0);
+        std::uniform_real_distribution<float> coord_dist(0.0, 1.0);
+        std::generate(vals_begin, vals_begin + vals_n, [&gen, &mass_dist, &velocity_dist, &coord_dist]() {
+            return Particle{mass_dist(gen), velocity_dist(gen), {coord_dist(gen), coord_dist(gen), coord_dist(gen)}};
+        });
+        std::transform(vals_begin, vals_begin + keys_n, keys_begin, [](const auto& particle) { return particle.energy(); });
+    }
+    else
+    {
+        TestUtils::generate_arithmetic_data(keys_begin, keys_n, seed);
+        TestUtils::generate_arithmetic_data(vals_begin, vals_n, seed + 1);
+        // avoid having value duplicates at the same locations as key duplicates for a more robust stability check
+        std::shuffle(vals_begin, vals_begin + keys_n, std::default_random_engine(seed));
+    }
+}
+
+template<typename Policy, typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_sort(Policy&& policy, KeyIt keys_begin, ValIt vals_begin, Size n, Compare compare, StableSortTag)
+{
+    if constexpr (std::is_same_v<Compare, NoComparatorTag>)
+        oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+    else
+        oneapi::dpl::stable_sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, compare);
+}
+
+template<typename Policy, typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_sort(Policy&& policy, KeyIt keys_begin, ValIt vals_begin, Size n, Compare compare, UnstableSortTag)
+{
+    if constexpr (std::is_same_v<Compare, NoComparatorTag>)
+        oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin);
+    else
+        oneapi::dpl::sort_by_key(policy, keys_begin, keys_begin + n, vals_begin, compare);
+}
+
+template<typename KeyIt, typename ValIt, typename Size, typename Compare>
+void
+call_reference_sort(KeyIt ref_keys_begin, ValIt ref_vals_begin, Size n, Compare)
```

Comment:
M.b. like that?

```
template<typename KeyIt, typename ValIt, typename Size, typename Compare = std::less<>>
void
call_reference_sort(KeyIt ref_keys_begin, ValIt ref_vals_begin, Size n, Compare comp = {})
{
    auto first = oneapi::dpl::make_zip_iterator(ref_keys_begin, ref_vals_begin);
    std::stable_sort(first, first + n, [](const auto& lhs, const auto& rhs) { return comp(std::get<0>(lhs), std::get<0>(rhs)); });
}

///....
call_reference_sort(....); //no comparator
call_reference_sort(...., GreaterComparatorTag{}); //with comparator
```

Reply:
I like the suggested idea for refactoring, but it would require changing the test in many other places, e.g. changing majority of functions to accept a variadic number of arguments, getting rid of `GreaterComparator` and `NoComparator` objects.

In my opinion, the test is in better shape right now, that it was before the PR. I suppose it may be treated as an "evolutionary" change. For more radical ones, let's stick to the idea from https://github.com/oneapi-src/oneDPL/pull/1692#discussion_r1701557694, and do it separately.

Reply:
I've created https://github.com/oneapi-src/oneDPL/issues/1780 for further refactoring.